### ISSUE TITLE
Split up default config.nu file

### ISSUE THREAD

**Issue by jmoore34** [nushell/nushell#4662]
### Related problem

It may be helpful to split up the default config.nu into multiple files, such as:

* prompt.nu
* completions/git.nu
* theme.nu

This could have the following advantages:

* Teaches users how to export and import modules
* Reduces the size of the config file to make it a little less intimidating
* In particular, if we write more git completions and the user wants to copy them, this way their `config.nu` doesn't become massive
* Makes it easier to share e.g. a color theme without worrying about overwriting or copying other parts of the config

It could also establish a convention for where to put external completions files. In the future, this could be nice:
* Installers for other CLI applications could place a completion file in this folder.
* A nu package manager could automatically place completion files in this folder.
* Perhaps eventually completion files could be parsed lazily, i.e., when the user calls the command in question. Then, if the user has 500 completion files for 500 different commands, they're not all parsed on startup.

### Describe the solution you'd like

* On first startup create a `completions` subdirectory in the config directory

* Split `config.nu` into `config.nu`, `completions/git.nu`, `theme.nu`, etc.

* For now, we could have `config.nu` just import from the other files. 

### Describe alternatives you've considered

Perhaps instead of creating `theme.nu`, we instead create `themes/default.nu` (might be nice for package manager integration / playing with different themes).

**Comment by fdncred** [comment]
ya, mine is like ~700 lines long. this seems reasonable but I'd like to relocate it to the `docs\sample_config` folder where it needs to be anyway (some !include_str code may have to changed when the file is moved). It would also be good to have a `custom-commands.nu` file that gets sourced as well and maybe `aliases.nu`, not sure exactly how far it needs to be broken out.

**Comment by jmoore34** [comment]
One issue I ran into here is that we can't currently `export` the `let theme = {...}`. Perhaps we could use `source` rather than `use` in the config.nu?

**Comment by jmoore34** [comment]
Also, I'm not sure which would be better stylistically:

a) Putting all the `source`s / `import`s at the top of `config.nu`

b) Calling `import` / `source` right before they're used

**Comment by fdncred** [comment]
i tend to put the source/use right before they're used.

semi-related to this is another feature we'd love to have at some point, which is a `nu-fmt` program to format nu scripts idomatically.

**Comment by jmoore34** [comment]
So I ran into a bit of an issue.

First, I tried doing `source completions/git.nu`, but this didn't work:

```
Error: nu::parser::file_not_found (https://docs.rs/nu-parser/0.59.1/nu-parser/enum.ParseError.html#variant.FileNotFound)

  × File not found
    ╭─[C:\Users\jon\AppData\Roaming\nushell\config.nu:54:1]
 54 │
 55 │ source "completions/git.nu"
    ·        ──────────┬─────────
    ·                  ╰── File not found: completions/git.nu
 56 │ #use "docs/sample_config/completions/git.nu" *
    ╰────
```

Changing it to an absolute path (or a path relative to the _current directory_) worked, though this isn't super practical.

I also tried changing to an `import`, which did accept a path relative to the `config.nu`; however, I got this error:

```
Error: nu::parser::parse_mismatch (https://docs.rs/nu-parser/0.59.1/nu-parser/enum.ParseError.html#variant.Expected)

  × Parse mismatch during operation.
   ╭─[docs/sample_config/completions/git.nu:1:1]
 1 │ export extern "git checkout" [
   ·        ───┬──
   ·           ╰── expected def or env keyword
 2 │   branch?: string@"nu-complete git branches" # name of the branch to checkout
   ╰────
```

(The same issue happens when trying to export `$config`)

**Comment by fdncred** [comment]
@jmoore34 we really need to add the `lib_dirs` functionality from 0.44 prior to our 0.60 release. this will enable us to put folder in that variable and, from any directory, just type source foo.nu and it would source it. I'm thinking such a thing would help this.

**Comment by fgimian** [comment]
Using `use` and `source` with a dynamic path seems to be troublesome here:

```
C:\Users\Fots〉use ($nu.config-path | path dirname | path join completions 'git.nu') *                           09/22/2022 07:54:52 PMError: nu::parser::module_not_found (link)

  × Module not found.
   ╭─[entry #45:1:1]
 1 │ use ($nu.config-path | path dirname | path join completions 'git.nu') *
   ·     ────────────────────────────────┬────────────────────────────────
   ·                                     ╰── module not found
   ╰────
  help: module files and their paths must be available before your script is run as parsing occurs before anything is evaluated

C:\Users\Fots〉source ($nu.config-path | path dirname | path join theme.nu)                                      09/22/2022 07:54:57 PMError: nu::parser::sourced_file_not_found (link)

  × File not found
   ╭─[entry #46:1:1]
 1 │ source ($nu.config-path | path dirname | path join theme.nu)
   ·        ──────────────────────────┬──────────────────────────
   ·                                  ╰── File not found: ($nu.config-path | path dirname | path join theme.nu)
   ╰────
  help: sourced files need to be available before your script is run
```

It seems the only way is to use an absolute path for each `use` or `source`, unless I'm missing something that is? 6599

**Comment by fdncred** [comment]
@fgimian there are several commands that do not support dynamic paths like `use` and `source`. That's actually what the help message is intended to convey. Perhaps we should be more explicit and just say, "You can't use variables to construct path for the use command".

**Comment by zoechi** [comment]
I was able to split it up by using static paths `use ~/.config/nushell/dark_theme.nu *` and `use` everywhere (no `source`) replacing variables to commands like

```
export def dark_theme [] {
  {
    # color for nushell primitives
    separator: white
    ....
  }
}
```
and then use it like `let-env config = { ... color_config: (dark_theme) ... }`

**Comment by dedebenui** [comment]
> "You can't use variables to construct path for the use command".

Is there any other way to have this functionality ? I'm trying to setup a separate file `my_env.nu` that sets a bunch of environment variables. Ideally I would want to simply put that file in `<nushell-config-dir>/scripts/my_env.nu` and call `use my_env.nu` in `env.nu`, but that doesn't work because the `$NU_LIB_DIRS` variable is not set before starting to parse `env.nu`. Other solutions are:

- write `use my_env.nu` in `config.nu` -> I guess that works but it's counterintuitive to set environment variable in `config.nu` while there is a `env.nu` explicitly for that purpose
- write `use /full-path-to/<nushell-config-dir>/scripts/my_env.nu` in `env.nu` -> that's not portable, I can no longer copy-paste my config from a machine to another machine.

Am I missing another solution ?

**Comment by dburriss** [comment]
Leaving a comment as maybe this is useful to someone for this. I have not tested this thoroughly but it seems to be working in that config (including theme) is overridden and aliases are set.

I am running version 0.83.0 on Mac OS.

The approach I use merges with the existing config with minimal changes. All you need to do is add the following `use` at the bottom of the default config.

```nushell
# config file found at $nu.config-path
# at the bottom of the file I place this
use ~/.config/nu/envs.nu
use ~/.config/nu/config.nu
use ~/.config/nu/theme.nu
use ~/.config/nu/aliases.nu
```

The _envs.nu_ file makes use of the `load-env` to load the record in, and then exports it using `export-env`. This will create an env var called `nu_env_var`.

```nushell
# ~/.config/nu/envs.nu
export-env { load-env {
    nu_env_var: "test"
}}
```

Then config uses the same method as above but now we are using `merge` to keep the existing set config values and overriding where needed.

```nushell
# ~/.config/nu/config.nu
export-env { load-env {
    config: ($env.config? | default {} | merge {
        show_banner: false
        edit_mode: vi
    })
}}
```

The theme carries on the same pattern as above but going into config and then color_config to set the theme. I just copied values from the doc page listed and it ov

```nushell
# ~/.config/nu/theme.nu
# example taken from https://www.nushell.sh/book/coloring_and_theming.html
export-env { load-env {
    config: ($env.config? | default {} | merge {
        color_config: ($env.config.color_config | default {} | merge {
            separator: "#88b719" # this sets only the foreground color like PR #486
            leading_trailing_space_bg: white # this sets only the foreground color in the original style
            header: { # this is like PR #489
                fg: "#B01455", # note, quotes are required on the values with hex colors
                bg: "#ffb900", # note, commas are not required, it could also be all on one line
                attr: bli # note, there are no quotes around this value. it works with or without quotes
            }
            date: "#75507B"
            filesize: "#729fcf"
            row_index: {
                # note, that this is another way to set only the foreground, no need to specify bg and attr
                fg: "#e50914"
            }
        })
    })
}}
```

```nushell
# ~/.config/nu/aliases.nu
alias uls = uutils-ls
```
This isn't pretty but I wanted to show the theming is definitely working.
<img width="527" alt="image" src="https://github.com/nushell/nushell/assets/5071495/d18b8b6d-c91f-4ab4-af76-be77cea7a9bf">

**Comment by MrPandir** [comment]
I'll add how I split my configuration, it might give you some ideas.
To some, it may show overengineering. 😅
<details><summary>Configuration</summary>
<p>

### config.nu
```nu
const name_theme_file = "default_dark.nu"

const dir_keybindings = ($nu.default-config-dir | path join keybindings)
mut keybindings = []
$keybindings ++= (source ($dir_keybindings | path join clipboard.nu))
$keybindings ++= (source ($dir_keybindings | path join control.nu))
$keybindings ++= (source ($dir_keybindings | path join cut.nu))
$keybindings ++= (source ($dir_keybindings | path join delete.nu))
$keybindings ++= (source ($dir_keybindings | path join menu.nu))
$keybindings ++= (source ($dir_keybindings | path join move.nu))
$keybindings ++= (source ($dir_keybindings | path join transformation.nu))

let menus = source ($nu.default-config-dir | path join menus.nu)

$env.config = {
    color_config: (source $name_theme_file)
    keybindings: $keybindings
    menus: $menus
	...     <- all other fields
}

# Autocompletes
use _git.nu *
...
use _docker.nu *
```

### env.nu
```nu
use prompts.nu

...       <- default ENV_CONVERSIONS and NU_PLUGIN_DIRS

$env.NU_LIB_DIRS = [
    ($nu.default-config-dir)
    ($nu.data-dir | path join "completions")
    ($nu.data-dir | path join "modules")
    ($nu.data-dir | path join "themes")
]
```

### promts.nu
```nu

def left_prompt [] {
    ...
}

def indicator [] {
    ...
}

def multiline_indicator [] {
    ...
}

export-env {
    $env.PROMPT_COMMAND = {|| left_prompt }
    $env.PROMPT_COMMAND_RIGHT = {|| null }

    $env.PROMPT_INDICATOR = {|| indicator }
    $env.PROMPT_MULTILINE_INDICATOR = {|| multiline_indicator }

    $env.PROMPT_INDICATOR_VI_INSERT = {|| indicator }
    $env.PROMPT_INDICATOR_VI_NORMAL = {|| indicator }
}

```

### /keybindings/control.nu
```nu
[
{
    name: _escape
    modifier: none
    keycode: escape
    mode: [emacs, vi_normal, vi_insert]
    event: { send: esc } 
}
{
    name: _cancel_command
    modifier: control
    keycode: char_c
    mode: [emacs, vi_normal, vi_insert]
    event: { send: ctrlc }
}
...
]
```
This is all divided into multiple files by functionality for convenience: `clipboard.nu`, `control.nu`, `cut.nu`, `delete.nu`, `menu.nu`, `move.nu`, `transformation.nu`.

### menus.nu
This is exactly the same list of `$env.config.menus` placed in this file. Contains `list`.

### ~/.local/share/nushell/
```
~/.local/share/nushell/
├── completions/
│   ├── _less.nu
│   └── _git.nu
├── modules/
│   └── run_script.nu
└── themes/
    ├── default_dark.nu   <- contains `record` with colors
    └── default_light.nu  <- contains `record` with colors
```

</p>
</details>

**Comment by NotTheDr01ds** [comment]
@fdncred It seems to me that pretty much every feature discussed here is now in Nushell:

- `env.nu`, `config.nu`, and vendor autoloads
- Ability to have modularized config parts rather than one monolithic, 700-line (850+ now) config record
- NU_LIB_DIRS

I think this case be closed?
